test/e2e/node/pre_stop.go

```diff
@@ -219,6 +219,67 @@ var _ = SIGDescribe("PreStop", func() {
 			return false, err
 		})
 	})
+
+	It("should block the Pod get terminated until preStop process finishes within termination grace period", func() {
+		gracefulTerminationPeriodSeconds := int64(60)
+		preStopProcessSec := 30
+		start := time.Now()
+		preStopFinish := start.Add(time.Second * time.Duration(preStopProcessSec))
+
+		By("creating the pod")
+		name := "pod-prestop-hook-" + string(uuid.NewUUID())
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{Name: name},
+			Spec: v1.PodSpec{
+				RestartPolicy: v1.RestartPolicyNever,
+				Containers: []v1.Container{
+					{
+						Name:    name,
+						Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+						Command: []string{"sh", "-c", "sleep 10"},
```

Comment:
This 10 needs to be longer. PreStop doesn't block a container from exiting or crashing. If the main process exits, that's game over.

Reply:
Thanks for bringing clarity @mikedanese. A User must ensure that preStop process MUST complete before the main process finishes. As per [document](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks) -
>It is blocking, meaning it is synchronous, so it must complete before the call to delete the container can be sent.

I need to change behavior of this test because above doc statement made me thinking unless preStop is finished, container status must not change and so the Pod status.

Reply:
Here is how I would simulate this test.
- Command running time should always be sufficiently larger than  preStop time. (say 5mins)
- if preStop time > gracefulTime then check if container state is not set to terminated until the prestop time is hit. (because the graceFulTime is increased in the increment of 2secs and until prestop is finished - this is what it means by blocking).
- if prestop time is < gracefultime then it gracefully ends.